examples/rcn/preproc.py

```diff
@@ -0,0 +1,171 @@
+"""A pre-processing layer of the RCN model. See Sec S8.1 for details.
+"""
+import logging
+
+import numpy as np
+from scipy.ndimage import maximum_filter
+from scipy.ndimage.filters import gaussian_filter
+from scipy.signal import fftconvolve
+
+LOG = logging.getLogger(__name__)
+
+
+class Preproc(object):
+    """
+    A simplified preprocessing layer implementing Gabor filters and suppression.
+
+    Parameters
+    ----------
+    num_orients : int
+        Number of edge filter orientations (over 2pi).
+    filter_scale : float
+        A scale parameter for the filters.
+    cross_channel_pooling : bool
+        Whether to pool across neighboring orientation channels (cf. Sec S8.1.4).
+
+    Attributes
+    ----------
+    filters : [numpy.ndarray]
+        Kernels for oriented Gabor filters.
+    pos_filters : [numpy.ndarray]
+        Kernels for oriented Gabor filters with all-positive values.
+    suppression_masks : numpy.ndarray
+        Masks for oriented non-max suppression.
+    """
+
+    def __init__(self, num_orients=16, filter_scale=4.0, cross_channel_pooling=False):
+        self.num_orients = num_orients
+        self.filter_scale = filter_scale
+        self.cross_channel_pooling = cross_channel_pooling
+        self.suppression_masks = generate_suppression_masks(
+            filter_scale=filter_scale, num_orients=num_orients
+        )
+
+    def fwd_infer(self, img, brightness_diff_threshold=40.0):
```

Comment:
Would be best to extract the relevant bits for this function and get rid of the redundant bits, and also move this into the standalone notebook. Directly copying this bit can be confusing for a tutorial-type example (especially when certain bits seem redundant).